### PR TITLE
Repository Scaffolding

### PR DESCRIPTION
**Pull request by jjgriff93**
Adds
- Devcontainer set-up
- Basic terraform and terragrunt set-up
- Linting
- Basic GH actions
- Makefile and READMEs

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ARG USERNAME=vscode
+ARG USER_UID=1000
+ARG USER_GID=$USER_UID
+RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+    && bash /tmp/library-scripts/common-debian.sh "${INSTALL_ZSH}" "${USERNAME}" "${USER_UID}" "${USER_GID}" "${UPGRADE_PACKAGES}" "true" "true" \ 
+    && apt-get install -y graphviz --no-install-recommends \
+    && apt-get clean -y && rm -rf /var/lib/apt/lists/*
+
+# [Option] Install Docker CLI
+ARG INSTALL_DOCKER="true"
```
Comment (t-young31) [inline .devcontainer/Dockerfile:34]: I'd be more opinionated and hard code as true but 🤷🏼

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+				"timonwong.shellcheck"
+			]
+		}
+	},
+
+	// Use 'forwardPorts' to make a list of ports inside the container available locally.
+	// "forwardPorts": [],
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	"postCreateCommand": "git config --global --add safe.directory /workspaces/FlowEHR",
```
Comment (t-young31) [inline .devcontainer/devcontainer.json:30]:
```suggestion
	"postCreateCommand": "git config --global --add safe.directory /workspaces/FlowEHR && cd /workspaces/FlowEHR && pre-commit install",
```
?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+deploy-transform: az-login ## Deploy transform infrastructure
+	$(call target_title, "Deploy Transform") \
+	&& . ${MAKEFILE_DIR}/scripts/load_env.sh \
+	&& terragrunt apply ${MAKEFILE_DIR}/infrastructure/transform
+
+deploy-serve: az-login ## Deploy serve infrastructure
+	$(call target_title, "Deploy Serve") \
+	&& . ${MAKEFILE_DIR}/scripts/load_env.sh \
+	&& terragrunt apply ${MAKEFILE_DIR}/infrastructure/serve
+
```
Comment (t-young31) [inline Makefile:69]: does `make deploy-serve && make destroy-all` work okay?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+destroy-all: az-login ## Destroy all infrastructure
```
Comment (t-young31) [inline Makefile:70]: maybe worth renaming to 
```suggestion
destroy: az-login ## Destroy all infrastructure
```
for consistency?

File: README.md [hunk 5]
```diff
@@ -1,2 +1,74 @@
...
+```bash
+cp config.sample.yaml config.yaml
+```
+
+Then edit `config.yaml` and specify the following values:
+
+- `prefix` - a prefix (max length 4 chars) to apply to all deployed resources (i.e. `flwr`)
+- `environment` - a unique name for your environment (i.e. `jgdev`)
+- `location` - the [Azure region](https://azuretracks.com/2021/04/current-azure-region-names-reference/) you wish to deploy resources to
+- `arm_subscription_id` - the [Azure subscription id](https://learn.microsoft.com/en-us/azure/azure-portal/get-subscription-tenant-id) you wish to deploy to
```
Comment (t-young31) [inline README.md:32]: I'm a little bit scared of having the doc for these scattered around (GH workflows, yaml comments, variables.tf, config_schema.json, README.md) but perhaps unavoidable?

File: infrastructure/core/main.tf [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  name                        = "${var.prefix}-${var.environment}-kv"
+  location                    = azurerm_resource_group.core.location
+  resource_group_name         = azurerm_resource_group.core.name
+  enabled_for_disk_encryption = true
+  tenant_id                   = data.azurerm_client_config.current.tenant_id
+  soft_delete_retention_days  = 7
+  purge_protection_enabled    = false
+  sku_name                    = "standard"
+  tags                        = var.tags
+
```
Comment (t-young31) [inline infrastructure/core/main.tf:62]: general question: when do we need

```hcl
  public_network_access_enabled   = false
```

File: scripts/load_env.sh [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  # Export as Terraform keys env vars
+  # shellcheck disable=SC2046
+  export $(yq e "$GET_LEAF_KEYS|$TF_KEYS| $FORMAT_FOR_ENV_EXPORT" config.yaml)
+
+  # Export additional required variables
+  export MGMT_RG="${PREFIX}-${ENVIRONMENT}-rg-mgmt"
+  export MGMT_ACR="${PREFIX}${ENVIRONMENT}acrmgmt"
+  export MGMT_STORAGE="${PREFIX}${ENVIRONMENT}strgmgmt"
+  export STATE_CONTAINER="tfstate"
+fi
```
Comment (t-young31) [inline scripts/load_env.sh:63]:
```suggestion
fi

# Export additional required variables
export MGMT_RG="${PREFIX}-${ENVIRONMENT}-rg-mgmt"
export MGMT_ACR="${PREFIX}${ENVIRONMENT}acrmgmt"
export MGMT_STORAGE="${PREFIX}${ENVIRONMENT}strgmgmt"
export STATE_CONTAINER="tfstate"
```
should always be exported(?)

File: scripts/remove_stale_ci_builds.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
Comment (t-young31) [inline scripts/remove_stale_ci_builds.sh:1]: probably should delete this file for the time being (apologies for adding it)

**Review by t-young31 — APPROVED** [review]
looks awesome to me – left a few small comments 😄

File: README.md (hunk 5 above)
Comment (jjgriff93) [reply README.md:32]: Very true - perhaps I leave these basics in but then say - for all other configuration options, see the schema

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (jjgriff93) [reply .devcontainer/Dockerfile:34]: Yeah, given we're not using docker in docker yet I'll leave for now as its from the base devcontainer reference, but agreed will probably remove in future

File: .devcontainer/devcontainer.json (hunk 2 above)
Comment (jjgriff93) [reply .devcontainer/devcontainer.json:30]: Great shout

File: scripts/remove_stale_ci_builds.sh (hunk 8 above)
Comment (jjgriff93) [reply scripts/remove_stale_ci_builds.sh:1]: Yep

File: infrastructure/core/main.tf (hunk 6 above)
Comment (jjgriff93) [reply infrastructure/core/main.tf:62]: Good question - likely when I've locked down the other resources as part of the terraform story. We'll enable purge protection at some point as well but want to pick my battles :D

File: Makefile (hunk 3 above)
Comment (jjgriff93) [reply Makefile:70]: They do now XD

**Pull request merged**